azurerm/resource_arm_iothub_resource_provider.go

```diff
@@ -0,0 +1,261 @@
+								string(devices.Free),
+								string(devices.Standard),
+							}, true),
+						},
+
+						"capacity": {
+							Type:         schema.TypeInt,
+							Required:     true,
+							ValidateFunc: validation.IntAtLeast(1),
+						},
+					},
+				},
+			},
+
+			"type": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"etag": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+
+			"hostname": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"shared_access_policy": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"key_name": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"primary_key": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"secondary_key": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"permissions": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+			},
+		},
+	}
+
+}
+
+func resourceArmIotHubCreateAndUpdate(d *schema.ResourceData, meta interface{}) error {
+
+	armClient := meta.(*ArmClient)
+	iothubClient := armClient.iothubResourceClient
+	ctx := armClient.StopContext
+
+	rg := d.Get("resource_group_name").(string)
+	name := d.Get("name").(string)
+
+	res, err := iothubClient.CheckNameAvailability(ctx, devices.OperationInputs{
+		Name: &name,
+	})
+
+	if err != nil {
+		return err
+	}
+
+	if !*res.NameAvailable {
+		return errors.New(string(res.Reason))
+	}
+
+	location := d.Get("location").(string)
+
+	subscriptionID := armClient.subscriptionId
+	skuInfo := expandAzureRmIotHubSku(d)
+
+	desc := devices.IotHubDescription{
+		Resourcegroup:  &rg,
+		Name:           &name,
+		Location:       &location,
+		Subscriptionid: &subscriptionID,
+		Sku:            &skuInfo,
+	}
+
+	if tagsI, ok := d.GetOk("tags"); ok {
+		tags := tagsI.(map[string]interface{})
+		desc.Tags = *expandTags(tags)
+	}
+
+	match := ""
+
+	if etagI, ok := d.GetOk("etag"); ok {
+		etag := etagI.(string)
+		desc.Etag = &etag
+		match = etag
+	}
+
+	future, err := iothubClient.CreateOrUpdate(ctx, rg, name, desc, match)
+	if err != nil {
+		return err
+	}
+
+	err = future.WaitForCompletion(ctx, iothubClient.Client)
+	if err != nil {
+		return fmt.Errorf("Error creating or updating IotHub %q (Resource Group %q): %+v", name, rg, err)
+	}
+
+	desc, err = iothubClient.Get(ctx, rg, name)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(*desc.ID)
+	return resourceArmIotHubRead(d, meta)
+
+}
+
+func expandAzureRmIotHubSku(d *schema.ResourceData) devices.IotHubSkuInfo {
+	skuList := d.Get("sku").([]interface{})
+	skuMap := skuList[0].(map[string]interface{})
+	cap := int64(skuMap["capacity"].(int))
+
+	return devices.IotHubSkuInfo{
+		Name:     devices.IotHubSku(skuMap["name"].(string)),
+		Tier:     devices.IotHubSkuTier(skuMap["tier"].(string)),
```

Comment:
_minor_ could we pull `name` and `tier` out above and cast them to a string there, since it'll make a crash log easier to understand, e.g.:

```
name := skuMap["name"].(string)
tier  := skuMap["tier"].(string)
# ...

return devices.IotHubSkuInfo{
  Name:     devices.IotHubSku(name),
  Tier:     devices.IotHubSkuTier(tier),
}

```